app/src/main/java/net/gsantner/markor/util/AppSettings.java

```diff
@@ -325,10 +325,69 @@ public void addRecentDocument(File file) {
         ShortcutUtils.setShortcuts(_context);
     }
 
+    public void copyFavouriteFile(File fileBefore, File fileAfter) {
+        List<String> list = new ArrayList<>();
+        List<File> favourites = getFavouriteFiles();
+        for (File f : favourites) {
+            // Only existing files or file which is 'virtual storage' is considered
+            if (f != null && (f.exists() || FilesystemViewerAdapter.isVirtualStorage(f))) {
+                list.add(f.getAbsolutePath());
+            }
+        }
+        String abs = fileBefore.getAbsolutePath();
+        boolean starred = false;
+        if (list.contains(abs)) {
+            starred = true;
+        }
+        if (starred) {
+            abs = fileAfter.getAbsolutePath();
+            if (!list.contains(abs)) {
+                list.add(abs);
+            }
+        }
+        setStringList(R.string.pref_key__favourite_files, list);
+    }
+
+    public void deleteFavouriteFile(File file) {
+        List<String> list = new ArrayList<>();
+        List<File> favourites = getFavouriteFilesAll();
+        String abs = file.getAbsolutePath();
+        for (File f : favourites) {
+            list.add(f.getAbsolutePath());
+        }
+        if (list.contains(abs)) {
+            list.remove(abs);
+        }
+        setStringList(R.string.pref_key__favourite_files, list);
+    }
+
+    public void renameFavoriteFile(File file, File newFile) {
```

Comment:
rename is a combination of copy (new file) and delete. So I think it's not needed to have a function for that

Right?

Reply:
The part dealing with rename is quite different from copy & delete. When copying & deleting, the implementation in original code would search down all sub-directories, while with renaming some times only a single method  provided by andorid library .renameTo is called upon. So with copying & deleting, each time a file's change is applied the function is called upon and that single file's favourite status is changed, but with renaming each time the renameFavourite function is called upon it might be dealing with a set of files(including all files in subdirectory which were starred before).